### PR TITLE
:bug: Valid domains won't validate

### LINKED ISSUE #10: Domains with hyphens are broken in 1.0.2
In fixing one issue with hyphens, I created another.

Domains like has-hyphen.example.com will fail with a stacktrace like this:
```
emailvalidator4j.parser.exception.DomainHyphen: Found -  in domain part
	at emailvalidator4j.parser.DomainPart.doParseDomainPart(DomainPart.java:59)
	at emailvalidator4j.parser.DomainPart.parse(DomainPart.java:39)
```

I'll follow up with a fix.

### PR DESCRIPTION
**Pull request by AlainODea**
- Allow valid domains with subdomains starting with numbers

 - Allow valid domains with hyphens

Resolves #10

### CODE REVIEW

File: src/main/java/emailvalidator4j/lexer/EmailLexer.java [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
     public void lex(String input) {
         Pattern pattern = Pattern.compile(
-                "(([a-zA-Z_]|[^\\u0000-\\u007F])+[46]?)|([0-9]+)|(\r\n)|(::)|(\\s+?)|(.)|(\\p{Cc}+)", Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE
+                "(([a-zA-Z0-9!#$%&'*+\\-/=?^_`{|}~]|[^\\u0000-\\u007F])+[46]?)|([0-9]+)|(\r\n)|(::)|(\\s+?)|(.)|(\\p{Cc}+)", Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE
```
Comment (AlainODea) [inline src/main/java/emailvalidator4j/lexer/EmailLexer.java]: Expand pattern to gather longest possible spans of atext into single GENERIC tokens. Follows atext as defined in [RFC 5322, Section 3.2.3](https://tools.ietf.org/html/rfc5322#section-3.2.3):
```plain
   atext           =   ALPHA / DIGIT /    ; Printable US-ASCII
                       "!" / "#" /        ;  characters not including
                       "$" / "%" /        ;  specials.  Used for atoms.
                       "&" / "'" /
                       "*" / "+" /
                       "-" / "/" /
                       "=" / "?" /
                       "^" / "_" /
                       "`" / "{" /
                       "|" / "}" /
                       "~"
```

and extended to UTF-8 by [RFC 6531, Section 3.3](https://tools.ietf.org/html/rfc6531#section-3.3):
```plain
atext   =/  UTF8-non-ascii
    ; extend the implicit definition of atext in
    ; RFC 5321, Section 4.1.2, which ultimately points to
    ; the actual definition in RFC 5322, Section 3.2.3
```

File: src/main/java/emailvalidator4j/parser/DomainPart.java [hunk 2]
```diff
@@ -12,6 +12,7 @@
 
 final class DomainPart extends Parser {
 
+    private static final Pattern subDomainIllegalCharacters = Pattern.compile("[^a-zA-Z0-9\\-\\x00007F-\\u10FFFF]");
```
Comment (AlainODea) [inline src/main/java/emailvalidator4j/parser/DomainPart.java]: Pick up any single character that is not printable ASCII or unicode.

File: src/main/java/emailvalidator4j/parser/DomainPart.java [hunk 3]
```diff
@@ -112,13 +113,25 @@ private boolean isDomainLiteral(boolean openBrackets) throws UnclosedDomainLiter
...
+        if (Tokens.GENERIC.equals(token.getName())) {
+            String subDomain = token.getText();
+            if (subDomain.charAt(0) == '-') {
+                throw new DomainHyphen("Sub-domains cannot start with hyphen");
+            }
+            if (subDomain.charAt(Math.max(0, subDomain.length() - 1)) == '-') {
+                throw new DomainHyphen("Sub-domains cannot end with hyphen");
+            }
+            Matcher matcher = subDomainIllegalCharacters.matcher(subDomain);
+            if (matcher.find()) {
```
Comment (AlainODea) [inline src/main/java/emailvalidator4j/parser/DomainPart.java]: Search for _any_ character that is not allowed in a domain name. One instance is fatal.

File: src/test/java/emailvalidator4j/EmailValidatorTest.java [hunk 4]
```diff
@@ -81,6 +81,10 @@ public void isInvalidEmail(String email) {
                 {"example((example))@fakedfake.co.uk"},
                 {"example@faked(fake).co.uk"},
                 {"example+@example.com"},
+                {"example@with-hyphen.example.com"},
+                {"with-hyphen@example.com"},
+                {"example@1leadingnum.example.com"},
+                {"1leadingnum@example.com"},
```
Comment (AlainODea) [inline src/test/java/emailvalidator4j/EmailValidatorTest.java]: Cover some of the email address I broke with my last change.

File: src/test/java/emailvalidator4j/lexer/EmailLexerTest.java [hunk 5]
```diff
@@ -167,6 +167,17 @@ public void toStringReturnsLeftTokens() {
...
     }
 
+    @Test
+    @UseDataProvider("atextExamples")
+    public void atextParsing(String atext) {
+        EmailLexer lexer = new EmailLexer();
+        lexer.lex(atext);
+        TokenInterface token = lexer.getCurrent();
+        Assert.assertEquals(Tokens.GENERIC, token.getName());
+        Assert.assertEquals(atext, token.getText());
```
Comment (AlainODea) [inline src/test/java/emailvalidator4j/lexer/EmailLexerTest.java]: Insist that atext be parsed as a single token even with mixed kinds of characters in it. Avoid parser regressions.

File: src/test/java/emailvalidator4j/lexer/EmailLexerTest.java [hunk 6]
```diff
@@ -178,6 +189,42 @@ public void invalidUTF8CharsAreInvalidTokens(String utf8String) throws Exception
...
+                {"a!"},
+                {"a\uD83D"},
+                {"aaa"},
+                {"a1a"},
+                {"a!a"},
+                {"a\uD83D!"},
+                {"aaa\uD83D"},
+                {"a1a\uD83D"},
+                {"a!a\uD83D"},
+                {"a\uD83Da\uD83D"},
```
Comment (AlainODea) [inline src/test/java/emailvalidator4j/lexer/EmailLexerTest.java]: Some trivial examples of mixed atext.

File: src/test/java/emailvalidator4j/parser/DomainPartTest.java [hunk 7]
```diff
@@ -54,7 +55,7 @@ public void invalidPartThrowsException(Class type, String domainPart) throws Inv
                 {DomainNotAllowedCharacter.class, "@example\\"},
                 {DomainNotAllowedCharacter.class, "@exa\\mple"},
                 {UnclosedDomainLiteral.class, "@example]"},
-                {ConsecutiveGeneric.class, "@example'"},
+                {DomainNotAllowedCharacter.class, "@example'"},
```
Comment (AlainODea) [inline src/test/java/emailvalidator4j/parser/DomainPartTest.java]: The failure mode here is more specific now. Consecutive generic was the best I could think of before and it wasn't very good.

File: src/test/java/emailvalidator4j/parser/DomainPartTest.java [hunk 8]
```diff
@@ -86,6 +87,8 @@ public void domainPartHasWarnings(String domainPart, List<Warnings> warnings) th
     @DataProvider
     public static Object[][] domainPartWithWarnings() {
         return new Object[][]{
+                {"@has-hyphen.example.com", Collections.emptyList()},
+                {"@1leadingnumber.example.com", Collections.emptyList()},
```
Comment (AlainODea) [inline src/test/java/emailvalidator4j/parser/DomainPartTest.java]: This is a bit odd. I'm asserting that these hyphenated and number-led domains parse without warnings. I'm possibly abusing this test, but it works.

**Comment by AlainODea** [comment]
@egulias if you get a chance to review this, I'd really appreciate it. I introduced a bad regression in the last release and it actually bit us in production. Sorry about that.

File: src/main/java/emailvalidator4j/parser/DomainPart.java [hunk 9]
```diff
@@ -112,13 +113,25 @@ private boolean isDomainLiteral(boolean openBrackets) throws UnclosedDomainLiter
...
 
-    private void checkNotAllowedChars(TokenInterface token) throws DomainNotAllowedCharacter {
+    private void checkNotAllowedChars(TokenInterface token) throws DomainNotAllowedCharacter, DomainHyphen {
         if (notAllowedTokens.contains(token)) {
             throw new DomainNotAllowedCharacter(
                     String.format("%s is not allowed in domain part", token.getName())
             );
         }
-
+        if (Tokens.GENERIC.equals(token.getName())) {
```
Comment (egulias) [inline src/main/java/emailvalidator4j/parser/DomainPart.java]: Maybe the whole `if` block can be extracted to `private void checkNotAllowedCharsInSubDomain` ?

File: src/test/java/emailvalidator4j/parser/DomainPartTest.java (hunk 8 above)
Comment (egulias) [reply src/test/java/emailvalidator4j/parser/DomainPartTest.java]: Yes, you are :). What do you think of creating two tests with this each case for now and if more examples are needed then creating a data provider?

File: src/main/java/emailvalidator4j/parser/DomainPart.java (hunk 9 above)
Comment (AlainODea) [reply src/main/java/emailvalidator4j/parser/DomainPart.java]: Good idea! I’ll do that.

File: src/test/java/emailvalidator4j/parser/DomainPartTest.java (hunk 8 above)
Comment (AlainODea) [reply src/test/java/emailvalidator4j/parser/DomainPartTest.java]: I agree. I’ll split these into a separate data provider and test for valid domain parts.

**Comment by AlainODea** [comment]
@egulias I think I have addressed your code review concerns and I'm happy to improve further as requested.

**Comment by AlainODea** [comment]
@egulias is this good to go? If I don't hear from you, I'll assume you are very busy and touch base next week.

**Comment by AlainODea** [comment]
@egulias any chance this can be reviewed and merged?

**Comment by egulias** [comment]
Hi @AlainODea , thanks for your patience. Yeas, busy weeks indeed. Merged. I'll do the release.

**Comment by AlainODea** [comment]
I sincerely appreciate this, @egulias. Thank you 😊

**Pull request merged**